test/ReverseProxy.FunctionalTests/WebSocketTests.cs

```diff
@@ -0,0 +1,241 @@
+                messageType,
+                endOfMessage: true,
+                cts.Token);
+
+            var message = await client.ReceiveAsync(buffer, cts.Token);
+
+            Assert.Equal(messageType, message.MessageType);
+            Assert.True(message.EndOfMessage);
+
+            var text = Encoding.UTF8.GetString(buffer.AsSpan(0, message.Count));
+            Assert.Equal(textToSend, text);
+
+            await client.CloseAsync(WebSocketCloseStatus.NormalClosure, "Bye", cts.Token);
+            Assert.Equal(WebSocketCloseStatus.NormalClosure, client.CloseStatus);
+            Assert.Equal("Bye", client.CloseStatusDescription);
+
+            await destinationHost.StopAsync(cts.Token);
+            await proxyHost.StopAsync(cts.Token);
+        }
+
+        [Fact]
+        public async Task RawUpgradeTest()
+        {
+            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
+
+            using var destinationHost = CreateWebSocketHost();
+            await destinationHost.StartAsync(cts.Token);
+            var destinationHostUrl = destinationHost.GetAddress();
+
+            using var proxyHost = CreateReverseProxyHost(destinationHostUrl);
+            await proxyHost.StartAsync(cts.Token);
+            var proxyHostUrl = proxyHost.GetAddress();
+
+            using var handler = new HttpClientHandler
+            {
+                AllowAutoRedirect = false,
+                AutomaticDecompression = DecompressionMethods.None,
+                UseCookies = false,
+                UseProxy = false
+            };
+            using var client = new HttpMessageInvoker(handler);
+            var webSocketsTarget = proxyHostUrl.Replace("https://", "wss://").Replace("http://", "ws://");
+            var targetUri = new Uri(new Uri(webSocketsTarget, UriKind.Absolute), "rawupgrade");
+            using var request = new HttpRequestMessage(HttpMethod.Get, targetUri);
+            request.Headers.TryAddWithoutValidation("Connection", "upgrade");
+            request.Version = new Version(1, 1);
+
+            var response = await client.SendAsync(request, cts.Token);
+
+            Assert.Equal(HttpStatusCode.SwitchingProtocols, response.StatusCode);
+
+#if NETCOREAPP3_1
+            using var rawStream = await response.Content.ReadAsStreamAsync();
+#elif NETCOREAPP5_0
+            using var rawStream = await response.Content.ReadAsStreamAsync(cts.Token);
+#else
+#error A target framework was added to the project and needs to be added to this condition.
+#endif
+
+            var buffer = new byte[1];
+            for (var i = 0; i <= 255; i++)
+            {
+                buffer[0] = (byte)i;
+                await rawStream.WriteAsync(buffer, 0, 1, cts.Token);
+                var read = await rawStream.ReadAsync(buffer, cts.Token);
+
+                Assert.Equal(1, read);
+                Assert.Equal(i, buffer[0]);
+            }
+
```

Comment:
The stream should get disposed before trying to stop the hosts, otherwise I think it would keep them from shutting down gracefully.
```suggestion

rawStream.Dispose();
```